src/main/java/io/reactivex/processors/AsyncProcessor.java

```diff
@@ -28,150 +26,244 @@
  * <p>The implementation of onXXX methods are technically thread-safe but non-serialized calls
  * to them may lead to undefined state in the currently subscribed Subscribers.
  * 
- * <p>Due to the nature Observables are constructed, the AsyncSubject can't be instantiated through
- * {@code new} but must be created via the {@link #create()} method.
- *
  * @param <T> the value type
  */
 public final class AsyncProcessor<T> extends FlowableProcessor<T> {
-    /** The state holding onto the latest value or error and the array of subscribers. */
-    final State<T> state;
-    /** 
-     * Indicates the subject has been terminated. It is checked in the onXXX methods in
-     * a relaxed matter: concurrent calls may not properly see it (which shouldn't happen if
-     * the reactive-streams contract is held).
-     */
-    boolean done;
+
+    @SuppressWarnings("rawtypes")
+    static final AsyncSubscription[] EMPTY = new AsyncSubscription[0];
+
+    @SuppressWarnings("rawtypes")
+    static final AsyncSubscription[] TERMINATED = new AsyncSubscription[0];
+
+    final AtomicReference<AsyncSubscription<T>[]> subscribers;
+    
+    /** Write before updating subscribers, read after reading subscribers as TERMINATED. */
+    Throwable error;
+    
+    /** Write before updating subscribers, read after reading subscribers as TERMINATED. */
+    T value;
     
     /**
-     * Constructs an empty AsyncSubject.
-     * @param <T> the observed and observable value type
-     * @return the new AsyncSubject instance.
+     * Creates a new AsyncProcessor.
+     * @param <T> the value type to be received and emitted
+     * @return the new AsyncProcessor instance
      */
     public static <T> AsyncProcessor<T> create() {
         return new AsyncProcessor<T>();
     }
     
-    protected AsyncProcessor() {
-        this.state = new State<T>();
+    /**
+     * Constructs an AsyncProcessor.
+     * @since 2.0
+     */
+    @SuppressWarnings("unchecked")
+    public AsyncProcessor() {
```

Comment:
`ReplaySubject(int, int)` makes no sense (in the current algorithm) because you are either unbounded (but want to avoid reallocating the buffer) or bounded (where you have a linked-list anyway). The benefit of the constructors may be more apparent in Java 7+ with diamond.